xcm/xcm-builder/src/barriers.rs

```diff
@@ -241,10 +241,18 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowExplicitUnpaidExecutionF
 			origin, instructions, max_weight, _weight_credit,
 		);
 		ensure!(T::contains(origin), ());
-		match instructions.first() {
-			Some(UnpaidExecution { weight_limit: Limited(m), .. }) if m.all_gte(max_weight) =>
+		// We will read up to 4 instructions. This allows up to 3 `DescendOrigin` instructions. We
+		// allow for more than one since anything beyond the first is a no-op and it's conceivable
```

Comment:
No it's not - you can descend as many levels as you like in principle (though practically speaking we must have a cutoff somewhere and 3 levels is probably fine for now).